File: crates/womp/src/lib/globalConfig.ts
import { invoke } from "@tauri-apps/api/core";
import { disable, enable, isEnabled } from "@tauri-apps/plugin-autostart";
import { create } from "zustand";

export type ThemeOption = "system" | "dark" | "light";

export interface GlobalConfig {
  autostart: boolean;
  theme: ThemeOption;
  tray_icon: string;
  run_commands: boolean;
  save_dpi_scale: boolean;
  save_icon_size: boolean;
  save_hdr_state: boolean;
  save_sdr_white_level: boolean;
  save_wallpaper_info: boolean;
  save_audio_output: boolean;
}

interface GlobalConfigStore {
  globalConfig: GlobalConfig;
  getGlobalConfig: () => Promise<GlobalConfig>;
  setGlobalConfig: (globalConfig: GlobalConfig) => void;
}

export const useGlobalConfigStore = create<GlobalConfigStore>((set) => ({
  globalConfig: {
    autostart: false,
    theme: "system",
    tray_icon: "womp",
    run_commands: true,
    save_dpi_scale: false,
    save_icon_size: false,
    save_hdr_state: false,
    save_sdr_white_level: false,
    save_wallpaper_info: false,
    save_audio_output: false,
  },
  getGlobalConfig: async () => {
    const globalConfig = await invoke<GlobalConfig>("get_global_config");
    set({ globalConfig });
    if (globalConfig.autostart) {
      if (!(await isEnabled())) {
        enable();
      }
    } else {
      if (await isEnabled()) {
        disable();
      }
    }
    return globalConfig;
  },
  setGlobalConfig: async (globalConfig: GlobalConfig) => {
    invoke("set_global_config", { globalConfig });
    set({ globalConfig });
    if (globalConfig.autostart) {
      if (!(await isEnabled())) {
        enable();
      }
    } else {
      if (await isEnabled()) {
        disable();
      }
    }
  },
}));

File: crates/womp-core/src/serde_types/global_config.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Default, Clone)]
pub struct GlobalConfig {
    pub autostart: bool,
    pub theme: String,
    pub tray_icon: String,
    pub run_commands: bool,
    pub save_dpi_scale: bool,
    pub save_icon_size: bool,
    pub save_hdr_state: bool,
    pub save_sdr_white_level: bool,
    pub save_wallpaper_info: bool,
    pub save_audio_output: bool,
}

impl GlobalConfig {
    pub fn new() -> Self {
        Self {
            autostart: false,
            theme: "system".to_string(),
            tray_icon: "womp".to_string(),
            run_commands: true,
            save_dpi_scale: false,
            save_icon_size: false,
            save_hdr_state: false,
            save_sdr_white_level: false,
            save_wallpaper_info: false,
            save_audio_output: false,
        }
    }
}
